### PR TITLE
Feature/cancel interaction

### PR DESCRIPTION
**Pull request by iCollin**
ATF Test Scripts to check [#2533](https://github.com/smartdevicelink/sdl_core/issues/2533)

This PR is **ready** for review.

### Summary
Add the ability to allow interaction alerts to be canceled.

### ATF version
develop

##### Enhancements
* Add an RPC called CancelInteraction that can dismiss any of an app's programmatically displayed alerts without needing to wait for the timeout to complete.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+--[[ Local functions ]]
+local function SendCancelInteracion()
+  local mobileSession = common.getMobileSession(1)
+  local hmiSession = common.getHMIConnection()
+  
+  local cid = mobileSession:SendRPC(rpcRequest.name, rpcRequest.params)
+  
+  EXPECT_HMICALL(rpcRequest.hmi_name, rpcRequest.params):Do(function(_, data)
+    hmiSession:SendResponse(data.id, data.method, "SUCCESS", {})
```
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: For the success case, the flow should be:

1. Mobile -> Core -> HMI: Slider()
2. Mobile -> Core -> HMI: CancelInteraction(SliderID, appID)
3. HMI -> Core -> Mobile: SliderResponse(ABORTED)
4. HMI -> Core -> Mobile: CancelInteractionResponse(SUCCESS)

[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: test cases are now realistic scenarios in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/48f2b20cf4294442ab524e732413ee0e655d00cd

File: test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local variables ]]
+local rpcRequest = {
+  name = "CancelInteraction",
+  hmi_name = "UI.CancelInteraction",
+  params = {
+    cancelID = -1,
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: -1 is replaced with the parameter being omitted in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/48f2b20cf4294442ab524e732413ee0e655d00cd

[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: additionally, a case was added with the parameter not matching the existing interaction

File: test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+--[[ Required Shared libraries ]]
+local runner = require('user_modules/script_runner')
+local common = require('user_modules/sequences/actions')
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local variables ]]
+local rpcInteraction = {
+  name = "Alert",
```
Comment (jacobkeeler) [inline test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua]: Add `cancelID`

File: test_scripts/API/CancelInteraction/003_SendCancelInteraction_ScrollableMessage_HMIReceivesRequest.lua [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local variables ]]
+local rpcInteraction = {
+  name = "ScrollableMessage",
+  hmi_name = "UI.ScrollableMessage",
+  params = {
+    scrollableMessageBody = "big scrollable\n\t\tmessage"
```
Comment (jacobkeeler) [inline test_scripts/API/CancelInteraction/003_SendCancelInteraction_ScrollableMessage_HMIReceivesRequest.lua]: Add `cancelID`

File: test_scripts/API/CancelInteraction/002_SendCancelInteraction_Alert_HMIReceivesRequest.lua [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local variables ]]
+local rpcInteraction = {
+  name = "Alert",
+  hmi_name = "UI.Alert",
+  params = {
+    alertText1 = "hello"
```
Comment (jacobkeeler) [inline test_scripts/API/CancelInteraction/002_SendCancelInteraction_Alert_HMIReceivesRequest.lua]: Add `cancelID` to this request and `CancelInteraction`

File: test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local variables ]]
+local rpcInteraction = {
+  name = "Slider",
+  hmi_name = "UI.Slider",
+  params = {
+    numTicks = 16,
```
Comment (jacobkeeler) [inline test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: Add `cancelID` to this request and `CancelInteraction`

File: test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: Should still include a separate test case where `cancelID` is omitted.

File: test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua [hunk 7]
```diff
@@ -31,18 +59,37 @@ local successResponse = {
...
   
-  local cid = mobileSession:SendRPC(rpcRequest.name, rpcRequest.params)
+  mobileSession:SendRPC(rpcCreateChoiceSet.name, rpcCreateChoiceSet.params)
+  local cid0 = mobileSession:SendRPC(rpcInteraction.name, rpcInteraction.params)
+  local cid1 = mobileSession:SendRPC(rpcCancelInteraction.name, rpcCancelInteraction.params)
   
-  EXPECT_HMICALL(rpcRequest.hmi_name, rpcRequest.params):Do(function(_, data)
+  EXPECT_HMICALL(rpcInteraction.hmi_name, {})
+  :Do(function(_, data)
+    hmiSession:SendResponse(data.id, data.method, "ABORTED", {})
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua]: fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+}
+
+local rpcInteraction = {
+  name = "PerformInteraction",
+  hmi_name = "UI.PerformInteraction",
+  hmi_name2 = "VR.PerformInteraction",
+  params = {
+    initialText = "hello",
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua]: fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+--[[ Local functions ]]
+local function SendCancelInteraction()
+  local mobileSession = common.getMobileSession(1)
+  local hmiSession = common.getHMIConnection()
+  
+  mobileSession:SendRPC(rpcCreateChoiceSet.name, rpcCreateChoiceSet.params)
+  local cid0 = mobileSession:SendRPC(rpcInteraction.name, rpcInteraction.params)
+  local cid1 = mobileSession:SendRPC(rpcCancelInteraction.name, rpcCancelInteraction.params)
+  
+  EXPECT_HMICALL(rpcInteraction.hmi_name, {})
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua]: fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730 , would have caught the scrollablemessage issue

File: test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    alertText1 = "hello"
+  }
+}
+
+local rpcCancelInteraction = {
+  name = "CancelInteraction",
+  hmi_name = "UI.CancelInteraction",
+  params = {
+    cancelID = 99,
+    functionID = 21489012
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua]: changed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+local function SendCancelInteraction()
+  local mobileSession = common.getMobileSession(1)
+  local hmiSession = common.getHMIConnection()
+  
+  local cid0 = mobileSession:SendRPC(rpcInteraction.name, rpcInteraction.params)
+  local cid1 = mobileSession:SendRPC(rpcCancelInteraction.name, rpcCancelInteraction.params)
+  
+  EXPECT_HMICALL(rpcInteraction.hmi_name, {})
+  :Do(function(_, data)
+    hmiSession:SendResponse(data.id, data.method, "SUCCESS", {})
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua]: fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+--  1) app1 is registered on SDL
+--
+--  Steps:
+--  1) app1 sends a sends an Alert RPC
+--  2) app1 sends a CancelInteraction Request with an INVALID functionID
+--
+--  Expected:
+--  1) the HMI does not receive the CancelInteraction request
+--  2) app1 receives INVALID_ID CancelInteraction response
+--  2) app1 receives SUCCESS from the Alert
```
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua]: fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua (hunk 3 above)
Comment (iCollin) [reply test_scripts/API/CancelInteraction/005_SendCancelInteraction_Other_HMIReceivesNoRequest.lua]: added in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/003_SendCancelInteraction_ScrollableMessage_HMIReceivesRequest.lua (hunk 4 above)
Comment (iCollin) [reply test_scripts/API/CancelInteraction/003_SendCancelInteraction_ScrollableMessage_HMIReceivesRequest.lua]: added in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/002_SendCancelInteraction_Alert_HMIReceivesRequest.lua (hunk 5 above)
Comment (iCollin) [reply test_scripts/API/CancelInteraction/002_SendCancelInteraction_Alert_HMIReceivesRequest.lua]: added to both in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730 , i added a seventh test case to omit cancelID

File: test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua (hunk 6 above)
Comment (iCollin) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: added to both in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua [hunk 13]
```diff
@@ -38,10 +37,32 @@ local rpcInteraction = {
...
+      { choiceID = 2, menuName = "choice2" }
+    },
+    initialText = {
+      fieldName = "initialInteractionText",
+      fieldText = "hello"
+    },
+    timeout = 10000
+  },
+  vr_params = {
+    helpPrompt = {
```
Comment (jacobkeeler) [inline test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua]: Add `cancelID = 99`

File: test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua [hunk 14]
```diff
@@ -53,16 +61,19 @@ local function SendCancelInteraction()
...
+  local interaction_id = 0
   
-  EXPECT_HMICALL(rpcInteraction.hmi_name, {})
+  EXPECT_HMICALL(rpcInteraction.hmi_name, rpcInteraction.hmi_params)
   :Do(function(_, data)
-    hmiSession:SendResponse(data.id, data.method, "SUCCESS", {})
+    interaction_id = data.id
   end)
 
+  local cid1 = mobileSession:SendRPC(rpcCancelInteraction.name, rpcCancelInteraction.params)
```
Comment (jacobkeeler) [inline test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua]: This should probably be in the `Do` block for 
```
EXPECT_HMICALL(rpcInteraction.hmi_name, rpcInteraction.hmi_params)
```
at least.

Comment (iCollin) [reply test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua]: moved in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/085ecf3126b192f8562f75bf9b3718c6cdde3111

File: test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua (hunk 13 above)
Comment (iCollin) [reply test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua]: added in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/085ecf3126b192f8562f75bf9b3718c6cdde3111

File: test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua (hunk 14 above)
Comment (iCollin) [reply test_scripts/API/CancelInteraction/006_SendCancelInteraction_WrongCancelID_HMIRespondsIgnored.lua]: i removed this. putting this line in the callback creates a race condition. I could use wait() but it isn't any better of a test so I'd rather not.

File: test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (iCollin) [reply test_scripts/API/CancelInteraction/004_SendCancelInteraction_Slider_HMIReceivesRequest.lua]: test case with cancelID omitted (007) added in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/471e8b5b7364595f4a5d0d0cb371f70f2d31a730

File: test_sets/SDL5_0/cancel_interaction.txt [hunk 15]
```diff
@@ -0,0 +1,7 @@
+./test_scripts/API/CancelInteraction/001_SendCancelInteraction_PerformInteraction_HMIReceivesRequest.lua
```
Comment (jacobkeeler) [inline test_sets/SDL5_0/cancel_interaction.txt]: This should be moved to `test_sets/cancel_interaction.txt`

Comment (iCollin) [reply test_sets/SDL5_0/cancel_interaction.txt]: moved in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2206/commits/d85311ff05a3e1d376c974df089d4949ad759edc

**Pull request merged**